sql/src/main/java/io/crate/metadata/Functions.java

```diff
@@ -79,19 +96,62 @@ public FunctionImplementation getSafe(FunctionIdent ident)
      */
     @Nullable
     public FunctionImplementation get(FunctionIdent ident) throws IllegalArgumentException {
-        FunctionImplementation implementation = functionImplementations.get(ident);
-        if (implementation != null) {
-            return implementation;
+        FunctionResolver dynamicResolver = functionResolvers.get(ident.name());
+        if (dynamicResolver != null) {
+            List<DataType> argumentTypes = ident.argumentTypes();
+            Signature signature = getMatchingSignature(argumentTypes, dynamicResolver);
+            if (signature != null) {
+                return dynamicResolver.getForTypes(replaceUndefinedDataTypesIfPossible(argumentTypes, signature));
+            }
         }
+        return null;
+    }
 
-        DynamicFunctionResolver dynamicResolver = functionResolvers.get(ident.name());
-        if (dynamicResolver != null) {
-            return dynamicResolver.getForTypes(ident.argumentTypes());
+    /**
+     * Iterate through list of possible {@link Signature} elements defined for the given {@link FunctionResolver}
+     * and return it if it matches the given list of data types.
+     */
+    @Nullable
+    private Signature getMatchingSignature(List<DataType> argumentTypes, FunctionResolver resolver) {
+        for (Signature signature : resolver.signatures()) {
+            if (signature.matches(argumentTypes)) {
+                return signature;
+            }
         }
         return null;
     }
 
     /**
+     * Iterator through list of given data types and replace possible {@link UndefinedType} elements with corresponding
+     * element at the matching signature. This will ensure, a function will be resolved with correct types.
+     *
+     * <p><strong>Note that if the corresponding element at the signature is an ANY type, it won't be used as a
+     * replacement for the {@link UndefinedType} as ANY types must only be used for signature matching.</strong></p>
+     */
+    private List<DataType> replaceUndefinedDataTypesIfPossible(List<DataType> argTypes, Signature matchingSignature) {
+        // we must create a new list here, as we can't be sure the incoming list is mutable
+        // TODO: discuss!
```

Comment:
seems like it the boolean impl outperforms the bitset variant:

```
Benchmark                                                           Mode  Cnt         Score         Error   Units
FunctionsBenchmark.benchReplaceUndefinedAlwaysNewList               thrpt  200  10004110.697 ±  108970.079   ops/s
FunctionsBenchmark.benchReplaceUndefinedAlwaysNewListNoReplacement  thrpt  200  11415906.498 ±  239981.813   ops/s
FunctionsBenchmark.benchReplaceUndefinedBitSet                      thrpt  200   8155821.527 ±  117787.736   ops/s
FunctionsBenchmark.benchReplaceUndefinedBitSetNoReplacement         thrpt  200  16351691.077 ±  506066.266   ops/s
FunctionsBenchmark.benchReplaceUndefinedBoolean                     thrpt  200   8316475.664 ±  124476.871   ops/s
FunctionsBenchmark.benchReplaceUndefinedBooleanNoReplacement        thrpt  200  19655331.352 ± 1039957.072   ops/s
```
https://gist.github.com/seut/a9362e1605451038ca30dcd3a8729cc1